langroid/agent/special/neo4j/csv_kg_chat.py

```diff
@@ -0,0 +1,139 @@
+
+    Note:
+    The function assumes the CSV file has a header row.
+    """
+    try:
+        df = read_csv(csv_location)
+        return df
+    except Exception as e:
+        raise Exception(f"Error occurred while reading the CSV file: {e}")
+
+
+def _preprocess_dataframe_for_neo4j(
+    df: DataFrame, default_value: Optional[str] = None, remove_null_rows: bool = True
+) -> DataFrame:
+    """
+    Preprocess a DataFrame for Neo4j import by fixing mismatched quotes in string
+        columns and handling null or missing values.
+
+    Args:
+        df (DataFrame): The DataFrame to be preprocessed.
+        default_value (str, optional): The default value to replace null values.
+        This is ignored if remove_null_rows is True. Defaults to None.
+        remove_null_rows (bool, optional): If True, rows with any null values will
+            be removed.
+        If False, null values will be filled with default_value. Defaults to False.
+
+    Returns:
+        DataFrame: The preprocessed DataFrame ready for Neo4j import.
+    """
+
+    # Fix mismatched quotes in string columns
+    for column in df.select_dtypes(include=["object"]):
+        df[column] = df[column].apply(
+            lambda x: x + '"' if (isinstance(x, str) and x.count('"') % 2 != 0) else x
+        )
+
+    # Handle null or missing values
+    if remove_null_rows:
+        df = df.dropna()
+    else:
+        if default_value is not None:
+            df = df.fillna(default_value)
+
+    return df
+
+
+class PandasToKGTool(ToolMessage):
+    request: str = "pandas_to_kg"
+    purpose: str = """Use this tool to create ONLY nodes and their relationships based
+    on the created model.
+    Take into account that the Cypher query will be executed while iterating 
+    over the rows in the CSV file (e.g. `index, row in df.iterrows()`),
+    so there NO NEED to load the CSV.
+    Make sure you send me the cypher query in this format: 
+    - placehoders in <cypherQuery> should be based on the CSV header. 
+    - <args> an array wherein each element corresponds to a placeholder in the 
+    <cypherQuery> and provided in the same order as the headers. 
+    SO the <args> should be the result of: `[row_dict[header] for header in headers]`
+    """
+    cypherQuery: str
+    args: list[str]
+
+    @classmethod
+    def examples(cls) -> List["ToolMessage"]:
+        return [
+            cls(
+                cypherQuery="""MERGE (employee:Employee {name: $employeeName, 
+                id: $employeeId})\n
+                MERGE (department:Department {name: $departmentName})\n
+                MERGE (employee)-[:WORKS_IN]->(department)\n
+                SET employee.email = $employeeEmail""",
+                args=["employeeName", "employeeId", "departmentName", "employeeEmail"],
+            ),
+        ]
+
+
+class CSVChatGraphAgent(Neo4jChatAgent):
+    def __init__(self, config: Neo4jChatAgentConfig):
```

Comment:
you may need to make a `CSVGraphChatAgentConfig`, so that the csv-chat specific `system_message` lives there, and it should be templated as usual with the provided csv schema. 

Also better name -> `CSVGraphChatAgent`, not `CSVChatGraphAgent`

Reply:
or just `CSVGraphAgent` and `CSVGraphAgentConfig` to keep it short.